core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -847,40 +612,37 @@ boolean isDeleteValueFinishNeeded(LookUp lookUp){
      * of concurrent removals since these removals will have to first call this function as well,
      * and they eventually change the version as well.
      *
-     * @param lookUp - It holds the entry to CAS, the previously written version of this entry
+     * @param ctx - It holds the entry to CAS, the previously written version of this entry
      *               and the value reference from which the correct version is read.
-     * @return a version is returned.
      *
      * If returned value (version) is {@code INVALID_VERSION} it means that a CAS was not preformed.
      * Otherwise, a positive version is returned, and it is the version written to the entry
      * (maybe by some other thread).
      * <p>
-     * Note 1: the version in the input param {@code lookUp} is updated in this method to be the
+     * Note 1: the version in the input param {@code ctx} is updated in this method to be the
      * updated one if a valid version was returned.
      *
      * Note 2: updating of the entries MUST be under published operation. The invoker of this method
      * is responsible to call it inside the publish/unpublish scope.
      */
-    int writeValueFinish(LookUp lookUp) { //TODO: check how not to return the version
-        int entryVersion = lookUp.version;
+    void writeValueFinish(ThreadContext ctx) {
+        final int entryVersion = ctx.value.getAllocVersion();
 
-        if (entryVersion > INVALID_VERSION) { // no need to complete a thing
-            return entryVersion;
+        if (entryVersion > Slice.INVALID_VERSION) { // no need to complete a thing
+            return;
         }
 
         // the value is not written yet or deleted (or in process of being deleted)
```

Comment:
It is not your fault but explanation here is a bit awkward. writeValueFinish() should be invoked in **only** one specific case, when reference is valid and version is invalid (or negative). So the version being invalid or negative is checked above (entryVersion > Slice.INVALID_VERSION), and if it is valid we return. And here the reference is checked if it id invalid we return. Can you please clarify it here or in the comment of the function header?

Reply:
fixed.